File: Cpp/54.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int rb = matrix[0].size()-1;
        int lob = matrix.size()-1;
        int lb = 0;
        int ub = 0;
        int dir = 1;
        int ctr = (int)matrix.size() * (int)matrix[0].size();
        vector<int> ans;
        int i=0,j=0;
        while(ctr--){
            ans.push_back(matrix[i][j]);
            switch(dir){
                case 1:
                    if(j == rb){
                        ub++;
                        dir = 2;
                        i++;
                    }else{
                        j++;
                    }
                    break;
                case 2:
                    if(i == lob){
                        rb--;
                        dir = 3;
                        j--;
                    }else{
                        i++;
                    }
                    break;
                case 3:
                    if(j == lb){
                        lob--;
                        dir = 4;
                        i--;
                    }else{
                        j--;
                    }
                    break;
                case 4:
                    if(i == ub){
                        lb++;
                        dir = 1;
                        j++;
                    }else{
                        i--;
                    }
                    break;
            }
        }
        return ans;
        
    }
};

File: Python/54.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rlimit = len(matrix[0])
        llimit = -1
        ulimit = -1
        dlimit = len(matrix)
        i = 0
        j = 0
        sol = []
        flag = True
        while(True):
            # go right
            flag = True
            while(j < rlimit):
                flag = False
                sol.append(matrix[i][j])
                j = j + 1
            if flag:
                break
            ulimit = ulimit + 1
            j = j - 1
            i = i + 1

            # go down
            flag = True
            while(i < dlimit):
                flag = False
                sol.append(matrix[i][j])
                i = i + 1
            if flag:
                break
            rlimit = rlimit - 1
            i = i - 1
            j = j - 1

            # go left
            flag = True
            while(j > llimit):
                flag = False
                sol.append(matrix[i][j])
                j = j - 1
            if flag:
                break
            
            dlimit = dlimit - 1
            j = j + 1
            i = i - 1

            # go up
            flag = True
            while(i > ulimit):
                flag = False
                sol.append(matrix[i][j])
                i = i - 1
            if flag:
                break
            
            llimit = llimit + 1
            i = i + 1
            j = j + 1
        return sol
